include/UtfView/code_unit_view.hpp

```diff
@@ -0,0 +1,608 @@
+// SPDX-License-Identifier: GPL-3.0-only
+
+#ifndef UTFVIEW_CODE_UNIT_VIEW_HPP
+#define UTFVIEW_CODE_UNIT_VIEW_HPP
+
+#include <UtfView/detail/concepts.hpp>
+#include <boost/stl_interfaces/iterator_interface.hpp>
+#include <iterator>
+#include <ranges>
+#include <type_traits>
+#include <utility>
+
+namespace utfview {
+
+/* PAPER: namespace std::uc { */
+
+/* PAPER */
+
+template <class I>
+consteval auto exposition_only_iterator_to_tag() { // @*exposition only*@
+  if constexpr (std::random_access_iterator<I>) {
+    return std::random_access_iterator_tag{};
+  } else if constexpr (std::bidirectional_iterator<I>) {
+    return std::bidirectional_iterator_tag{};
+  } else if constexpr (std::forward_iterator<I>) {
+    return std::forward_iterator_tag{};
+  } else if constexpr (std::input_iterator<I>) {
+    return std::input_iterator_tag{};
+  }
+}
+
+template <class I>
+using exposition_only_iterator_to_tag_t =
+    decltype(exposition_only_iterator_to_tag<I>()); // @*exposition only*@
+
+template <typename V, typename ToType>
+concept exposition_only_convertible_to_charN_t_view =
+    exposition_only_code_unit_to<ToType> && std::ranges::view<V> &&
+    std::convertible_to<std::ranges::range_reference_t<V>, ToType>;
+
+template <exposition_only_convertible_to_charN_t_view<char8_t> V>
+class as_char8_t_view : public std::ranges::view_interface<as_char8_t_view<V>> {
+  V base_ = V(); // @*exposition only*@
+
+  template <bool Const>
```

Comment:
Why are these exposition-only?  Most (all?) C++ views have nested `iterator` and `sentinel` templates.

Reply:
They have nested exposition-only *iterator* and *sentinel* types, as far as I can tell: https://en.cppreference.com/w/cpp/ranges/elements_view/iterator

Reply:
Ha!  So they are.  I think I just misremembered.

Reply:
Right, the correct way to get the types is `iterator_t` and `sentinel_t` which are just `decltype(ranges::begin(r))` and similarly for end. So the types don't need to have publicly visible names as nested types of the view.